### ISSUE TITLE
Improve response to scanning non-CWA lab QR code

### ISSUE THREAD

**Issue by MikeMcC399** [corona-warn-app/cwa-wishlist#343]
## Current Implementation

Some labs issue QR codes after Corona testing which use a system that is not compatible with CWA QA codes (see https://github.com/corona-warn-app/cwa-wishlist/issues/270#issuecomment-764986222).

If a user scans such a QR code the error message is correct, but it doesn't distinguish between a foreign QR code and a CWA QR code which has already been registered.

## Steps to reproduce

CWA Android version 1.10.1

1. Open CWA
2. Tap NEXT STEPS > Test with QR Code
3. ACCEPT (Your Consent)
4. Position the QR code in the frame<br/>This QR code contains the text "DE12345678"

![QR Code DE12345678](https://user-images.githubusercontent.com/66998419/105467552-25905d80-5c96-11eb-8408-b4835ee6d61f.png)
5. Note the error message shown after scanning
![QR Code is invalid](https://user-images.githubusercontent.com/66998419/105468344-2d043680-5c97-11eb-853f-0d6d7f2c291a.jpg)

## Suggested Enhancement

The scanned QR code should be checked locally in the app to determine whether it has the right format.

The format should be 
https://localhost/?XXXXXX-XXXXXXXX-XXXX-XXXX-XXXX-XXXXXXXXXXXX for instance
https://localhost/?3D6D08-3567F3F2-4DCF-43A3-8737-4CD1F87D6FDA

If the QR code does not correspond to this format, then the error message should say so, and should not say that the code might have been registered on another smartphone already.

A special case could also be made if the user somehow had access to the 10C/OECD lab form and scanned this in. The format of the the lab QR code is
XXXXXX-XXXXXXXX-XXXX-XXXX-XXXX-XXXXXXXXXXXX for instance
3D6D08-3567F3F2-4DCF-43A3-8737-4CD1F87D6FDA

If the QR code scanned in is not recognisable as a CWA QR code, then the app could display an example of a correct form (Muster 10C).

If the user determines that they have attempted to scan in a CWA QR code from the correct form, and the QR code is rejected due to formatting issues, the user should be advised where they can report this problem. 

## Expected Benefits

Clearer differentiation of error conditions leading to better user satisfaction and resolution.

**Comment by dsarkar** [comment]
@MikeMcC399 actually, what do you think about extending your proposal:
- Additionally, the notification should remind at that stage to get a teleTAN from the hotline.

**Comment by MikeMcC399** [comment]
@dsarkar 
> what do you think about extending your proposal:
>
>- Additionally, the notification should remind at that stage to get a teleTAN from the hotline.

That is not the right time to be proposing requesting a teleTAN. At the time of scanning a QR code, the user is registering a Corona test. They do not (normally) know yet if the test is positive. A teleTAN should only be requested if the user has **already received a positive test result and**:

1. has received a QR code for CWA and it didn't work or
2. has not been prompted on the device to share the result or
3. wants to report a positive test result on an additional device

So, no the proposal should not be extended.

**Comment by Ein-Tim** [comment]
I disagree. I already heard from some people that there are some labs who only issues the QR-Code when somebody is tested positive (yes, I know this is not how it is supposed to work - but I think the labs are doing this because they have too much work to do). If now the QR-Code the lab issued after the test was positive is not a valid CWA QR-Code, I don't think there would be a problem with adding (for example):

"If you know that you have tested positive for COVID-19, you can call the Verification Hotline to get a TeleTAN and warn other users."

**Comment by MikeMcC399** [comment]
@Ein-Tim / @dsarkar 
I take your comments on board. So yes, add the point to recommend requesting a TAN (in the app it is not called teleTAN) if the user has already received a positive test result.

**Comment by dsarkar** [comment]
Internal Tracking ID:  [EXPOSUREAPP-4721](https://jira-ibs.wbs.net.sap/browse/EXPOSUREAPP-4721)

Updates will be published here.

**Comment by Ein-Tim** [comment]
The error message has changed with version 1.12: https://github.com/corona-warn-app/cwa-app-ios/issues/1386#issuecomment-777437011

**Comment by dsarkar** [comment]
Actually, the message text was implemented earlier!

**Comment by MikeMcC399** [comment]
@Ein-Tim 

If you compare the message in the original post #343 with the one you referred to from https://github.com/corona-warn-app/cwa-app-ios/issues/1386#issuecomment-777437011 you will see that they are identical.

I did not research iOS, but for Android the message with id `submission_error_dialog_web_test_paired_body` has been in exactly this form since CWA Android 1.9.1. See https://github.com/corona-warn-app/cwa-app-android/blob/5bfc89c1d91bf8afd6aa610dbd611b4515d30e46/Corona-Warn-App/src/main/res/values/strings.xml#L753-L754

 <string name="submission_error_dialog_web_test_paired_body">"The QR code is invalid or has been registered on another smartphone already. You will receive your test result from the test center or laboratory regardless of the validity of the QR code. If you are diagnosed with coronavirus, you will be notified by the public health authority."</string>

The point about this wishlist item (#343) is to distinguish between a wrongly formatted QR code and a QR code which has already been used. The two different cases should cause two different and appropriate error messages to be displayed.

**Comment by Ein-Tim** [comment]
Okay, this was a bad one from me. 

Sorry, I mixed things up. 

I'm so sorry for the confusion. 😕

**Comment by Ein-Tim** [comment]
Related: https://github.com/corona-warn-app/cwa-documentation/issues/429

**Comment by dsarkar** [comment]
new ticket [EXPOSUREAPP-4406](https://jira-ibs.wbs.net.sap/browse/EXPOSUREAPP-4406)

**Comment by Manawyrm** [comment]
In lower saxony, regular QR codes with 8 character A-Z, 0-9 are present on the letters from the vaccination centers.
The regex `^[0-9A-Z]{8}$` matches against these NDS QR codes and could maybe used to detect them and warn the user? 

I've seen some confusion in people, because they tried to register with that code.

**Comment by MikeMcC399** [comment]
It looks like this issue will be addressed by CWA 2.11 in Android where there is a universal QR code scanner under development. The current CWA 2.11 Android branch under development outputs the error message "Kein geeigneter QR-Code" when scanning the following random QR code containing the text "DE12345678" which is not related to any CWA standard QR code:

![QR Code DE12345678](https://user-images.githubusercontent.com/66998419/105467552-25905d80-5c96-11eb-8408-b4835ee6d61f.png)

I'll look at this again when some more of the development work has been completed for 2.11.

**Comment by MikeMcC399** [comment]
I looked at the scan QR code function in the Android pre-release version 2.11.0-RC0 (commit https://github.com/corona-warn-app/cwa-app-android/commit/a79fb7b3e1e690b3d667906adb913f10cab14343) when scanning the non-CWA QR code "DE12345678" from the image https://user-images.githubusercontent.com/66998419/105467552-25905d80-5c96-11eb-8408-b4835ee6d61f.png displayed on another device.

The warning message which comes up says "No suitable QR code". That would fulfill the needs of this issue when the functionality is released.

There are some minor issues with the error text:

1. In the English translation "scanned" is misspelled as "canned"
2. In both English and German if a QR code is scanned using the camera, the error text suggests selecting a suitable file instead of selecting a different QR code. If the QR code was scanned via the camera then the file option is not necessarily relevant.

![No_suitable_QR_code](https://user-images.githubusercontent.com/66998419/134478769-d006b368-333a-4138-b3fa-53efb63dc21f.jpg)

The "DETAILS" option leads to a stack trace. This is not optimal as the outcome of a user error. If the "DETAILS" option led to the display of the QR code contents in text form this might help troubleshooting. Alternatively the error message could link to a new FAQ entry which explains in more detail which QR codes can be scanned.

![Scan_stack_trace](https://user-images.githubusercontent.com/66998419/134478810-a3da0e5e-53b0-41c3-b65a-fef98e98500e.jpg)

I don't know whether this feature is supposed to be finished or not. If yes, then I should probably submit issues into [cwa-app-android](https://github.com/corona-warn-app/cwa-app-android/issues), otherwise I will wait.

**Comment by MikeMcC399** [comment]
The original issue has been resolved with the Android 2.11 release.
Internal Tracking ID:  [EXPOSUREAPP-4406](https://jira-ibs.wbs.net.sap/browse/EXPOSUREAPP-4406)

Some more improvements are needed though and these are covered in the separate issues:
- #4155 "Typo in QR scan error message (EN"
- #4156 "Reference to file in QR scan from camera error message"
- #4158 "Stack trace after scanning incompatible QR code"

I will leave this open for the moment in case there is any feedback on the iOS 2.11 release in this area.

**Comment by Ein-Tim** [comment]
> I will leave this open for the moment in case there is any feedback on the iOS 2.11 release in this area.

Hello There 😉

iOS shows "The QR code is not supported by the Corona-Warn-App. Please scan a suitable QR code." when attempting to scan the QR code @MikeMcC399 shared in [his comment above](https://github.com/corona-warn-app/cwa-wishlist/issues/343#issuecomment-923863479).

From my point of view, this issue can be closed.

**Comment by MikeMcC399** [comment]
Closing.

Android issues being followed separately, see https://github.com/corona-warn-app/cwa-wishlist/issues/343#issuecomment-936580161.

**Comment by dsarkar** [comment]
@MikeMcC399 Thanks, I forwarded your comment https://github.com/corona-warn-app/cwa-wishlist/issues/343#issuecomment-936580161  into the original internal ticket. Good night!
